homeassistant/components/esphome/voice_assistant.py

```diff
@@ -0,0 +1,164 @@
+"""ESPHome voice assistant support."""
+from __future__ import annotations
+
+import asyncio
+from collections.abc import AsyncIterable, Callable
+import logging
+import socket
+from typing import cast
+
+from aioesphomeapi import VoiceAssistantEventType
+
+from homeassistant.components import stt
+from homeassistant.components.assist_pipeline import (
+    PipelineEvent,
+    PipelineEventType,
+    async_pipeline_from_audio_stream,
+)
+from homeassistant.components.media_player import async_process_play_media_url
+from homeassistant.core import HomeAssistant, callback
+
+from .enum_mapper import EsphomeEnumMapper
+
+_LOGGER = logging.getLogger(__name__)
+
+UDP_PORT = 0  # Set to 0 to let the OS pick a free random port
+
+_VOICE_ASSISTANT_EVENT_TYPES: EsphomeEnumMapper[
+    VoiceAssistantEventType, PipelineEventType
+] = EsphomeEnumMapper(
+    {
+        VoiceAssistantEventType.VOICE_ASSISTANT_ERROR: PipelineEventType.ERROR,
+        VoiceAssistantEventType.VOICE_ASSISTANT_RUN_START: PipelineEventType.RUN_START,
+        VoiceAssistantEventType.VOICE_ASSISTANT_RUN_END: PipelineEventType.RUN_END,
+        VoiceAssistantEventType.VOICE_ASSISTANT_STT_START: PipelineEventType.STT_START,
+        VoiceAssistantEventType.VOICE_ASSISTANT_STT_END: PipelineEventType.STT_END,
+        VoiceAssistantEventType.VOICE_ASSISTANT_INTENT_START: PipelineEventType.INTENT_START,
+        VoiceAssistantEventType.VOICE_ASSISTANT_INTENT_END: PipelineEventType.INTENT_END,
+        VoiceAssistantEventType.VOICE_ASSISTANT_TTS_START: PipelineEventType.TTS_START,
+        VoiceAssistantEventType.VOICE_ASSISTANT_TTS_END: PipelineEventType.TTS_END,
+    }
+)
+
+
+class VoiceAssistantUDPServer(asyncio.DatagramProtocol):
+    """Receive UDP packets and forward them to the voice assistant."""
+
+    started = False
+    queue: asyncio.Queue[bytes] | None = None
+    transport: asyncio.DatagramTransport | None = None
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize UDP receiver."""
+        self.hass = hass
+        self.queue = asyncio.Queue()
+
+    async def start_server(self) -> int:
+        """Start accepting connections."""
+
+        def accept_connection() -> VoiceAssistantUDPServer:
+            """Accept connection."""
+            if self.started:
+                raise RuntimeError("Can only start once")
+            if self.queue is None:
+                raise RuntimeError("No longer accepting connections")
+
+            self.started = True
+            return self
+
+        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
```

Comment:
Why isn't the aioesphomeapi library hosting the server code? The server could expose the `_iterate_packets` method for library users.